walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,135 @@
+
+    // Make the remaining free blocks continuous and verify they are of minimum duration
+    updateMeetingTimes(meetingTimes, freeTimes, meetingDuration);
+    
+    // Verify that there is at least one potential meeting time with all attendees
+    // If not, only consider mandatory attendees (if any)
+    if (meetingTimes.size() != 0) {
+      return meetingTimes;
+    }
+
+    // Consider only mandatory attendees
+    if (mandatoryAttendees.size() != 0) {
+      freeTimes.clear();
+      populateFreeTimes(freeTimes, 30);
+      removeBusyTimes(freeTimes, events, mandatoryAttendees, 30);
+      updateMeetingTimes(meetingTimes, freeTimes, meetingDuration);
+    }
+    return meetingTimes;
+  }
+
+  private void populateFreeTimes (ArrayList<TimeRange> freeTimes, int blockLength) {
+    /* Populates FreeTimes with 1 day of blocks of size blockLength
+     * (i.e. blocklength 15 would mean 12:00am, 12:15am, 12:30am, 12:45am, etc)
+     * Note that blockLength must be a factor of 60 minutes or there may be unexpected behaviour.
+     */
+
+    // blocks holds all possible minutes value in the hour (ex: 0, 15, 30 if blockLength=15)
+    ArrayList<Integer> blocks = new ArrayList<Integer>();
+    for (int i = 0; i < 60; i++) {
+      if (i % blockLength == 0){
+        blocks.add(i);
+      }
+    }
+
+    // Save all TimeRanges
+    for (int i = 0; i < 24; i++) {
+      for (int b : blocks){
+        int startTime = TimeRange.getTimeInMinutes(i, b);
+        freeTimes.add(TimeRange.fromStartDuration(startTime, blockLength));
+      }
+    }
+  }
+
+  private void removeBusyTimes(ArrayList<TimeRange> freeTimes, Collection<Event> events, Collection<String> meetingAttendees, int blockLength) {
+    /* Updates freeTimes in place by removing all TimeRanges which overlap with events where
+     * at least one meeting attendee is attending.
+     */
+    
+    // Consider all events
+    for (Event e : events){
+
+      // Check if the event has at least one meetingAttendee (otherwise just skip the event)
+      ArrayList<String> eventAttendees = new ArrayList<String>(e.getAttendees());
+      eventAttendees.retainAll(meetingAttendees);
+      if (eventAttendees.size() != 0) {
+
+        // Remove all busy time blocks (those in the event) from list of free times
+        TimeRange eventTime = e.getWhen();
+        int numBlocks = eventTime.duration() / blockLength;
+        for (int i = 0; i < numBlocks; i++) {
+          freeTimes.remove(TimeRange.fromStartDuration(eventTime.start() + i * blockLength, blockLength));
+        }
+      }
+    }
+  }
+
+  private void updateMeetingTimes(Collection<TimeRange> meetingTimes, ArrayList<TimeRange> freeTimes, long meetingDuration){
+    /* Update meetingTimes in place by adding blocks of uninterrupted freeTimes of minimum length meetingDuration.
+     * This is done by merging continuous blocks of freeTime into one big range.
+     */
+
+    // Merge consecutive times blocks
+    while (freeTimes.size() != 0) {
```

Comment:
this can be a for loop in which you act whenever you find freeTimes.get(i-1).end() != freeTimes.get(i).start()
- keep track of the range start and use it to build a range if exceeds meetingDuration
- update the range start to i.